tests/networkpolicy_test.go

```diff
@@ -1,173 +1,236 @@
 package tests_test
 
 import (
+	"fmt"
+	"net"
+	"strconv"
 	"time"
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
-	v1network "k8s.io/api/networking/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	networkv1 "k8s.io/api/networking/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
-	v13 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
 
 	v1 "kubevirt.io/client-go/api/v1"
 	"kubevirt.io/client-go/kubecli"
 	"kubevirt.io/kubevirt/tests"
 	cd "kubevirt.io/kubevirt/tests/containerdisk"
+	"kubevirt.io/kubevirt/tests/libvmi"
 )
 
 var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:component]Networkpolicy", func() {
 
-	var virtClient kubecli.KubevirtClient
-
-	var vmia *v1.VirtualMachineInstance
-	var vmib *v1.VirtualMachineInstance
-	var vmic *v1.VirtualMachineInstance
-
-	tests.BeforeAll(func() {
+	var (
+		virtClient                                          kubecli.KubevirtClient
+		serverVMI, clientVMI, clientVMIAlternativeNamespace *v1.VirtualMachineInstance
+		serverVMILabels                                     = map[string]string{"type": "test"}
+	)
+	BeforeEach(func() {
 		var err error
-
 		virtClient, err = kubecli.GetKubevirtClient()
-		tests.PanicOnError(err)
+		Expect(err).ToNot(HaveOccurred(), "should succeed retrieving the kubevit client")
 
 		tests.SkipIfUseFlannel(virtClient)
 		skipNetworkPolicyRunningOnKindInfra()
-		tests.BeforeTestCleanup()
-
-		// Create three vmis, vmia and vmib are in same namespace, vmic is in different namespace
-		vmia = createVMICirros(virtClient, tests.NamespaceTestDefault, map[string]string{"type": "test"})
-		vmib = createVMICirros(virtClient, tests.NamespaceTestDefault, map[string]string{})
-		vmic = createVMICirros(virtClient, tests.NamespaceTestAlternative, map[string]string{})
 
-		vmia = tests.WaitUntilVMIReady(vmia, tests.LoggedInCirrosExpecter)
-		vmib = tests.WaitUntilVMIReady(vmib, tests.LoggedInCirrosExpecter)
-		vmic = tests.WaitUntilVMIReady(vmic, tests.LoggedInCirrosExpecter)
 	})
+	Context("when three cirros VMs with default networking are started and serverVMI start an HTTP server on port 80 and 81", func() {
+		tests.BeforeAll(func() {
```

Comment:
Not sure we want to do that, but this could be abstracted further; your pattern is: 
  - generate vmi spec
  - instantiate vmi spec
  - wait for the vmi to be ready
  - assert the vmi has IPs.

You could create a function for that, e.g.

```golang

createReadyVMIWithIPs := func(vmi *v1.VirtualMachineInstance, serverCreationFunc, ports ...int) *v1.VirtualMachineInstance {
    readyVMI := tests.WaitUntilVMIReady(serverVMI, tests.LoggedInCirrosExpecter)
    
    for _, port := range ports {
        tests.StartServer(vmi, serverCreationFunc, port)
    }
    assertIPsNotEmptyForVMI(vmi)
}
```

It would be invoked like: 

```golang
serverVMI = createVMICirros(virtClient,
		tests.NamespaceTestDefault,
		serverVMILabels,
		libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding(
			v1.Port{
				Name:     "http80",
				Port:     80,
				Protocol: "TCP",
			},
			v1.Port{
				Name:     "http81",
				Port:     81,
				Protocol: "TCP",
			},
		)),
	)

createReadyVMIWithIPs(serverVMI, tests.HTTPServer, 80, 81)
```

Reply:
Edy raised in the past that it will cause the vms not to load in parallel
so lets leave the Create versus the wait in two phases?

Reply:
Yep that's why we do create create create wait wait wait at the networkpolicy_test we are not going to change that, thanks from point this out @oshoval

Reply:
Resolving this.

Reply:
Wait, why is this exclusive ? 

You can 

```golang
serverVMI = createVMICirros(virtClient,
		tests.NamespaceTestDefault,
		serverVMILabels,
		libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding(
			v1.Port{
				Name:     "http80",
				Port:     80,
				Protocol: "TCP",
			},
			v1.Port{
				Name:     "http81",
				Port:     81,
				Protocol: "TCP",
			},
		)),
	)
vmi2 = createVMICirros(virtClient,...)
vmi3 = createVMICirros(virtClient,...)

haveReadyVMIWithIPs(serverVMI, ...)
haveReadyVMIWithIPs(vmi2, ...)
haveReadyVMIWithIPs(vmi3, ...)
```